src/analyzer/analyzerrhythm.cpp

```diff
@@ -0,0 +1,228 @@
+#include "analyzer/analyzerrhythm.h"
+
+#include <QHash>
+#include <QString>
+#include <QVector>
+#include <QtDebug>
+
+#include "analyzer/constants.h"
+#include "track/beatfactory.h"
+#include "track/beatmap.h"
+#include "track/beatutils.h"
+#include "track/track.h"
+
+namespace {
+
+// This determines the resolution of the resulting BeatMap.
+// ~12 ms (86 Hz) is a fair compromise between accuracy and analysis speed,
+// also matching the preferred window/step sizes from BeatTrack VAMP.
+// For a 44.1 kHz track, we go in 512 sample steps
+// TODO: kStepSecs and the waveform sample rate of 441
+// (defined in AnalyzerWaveform::initialize) do not align well and thus
+// generate interference. Currently we are at this odd factor: 441 * 0.01161 = 5.12.
+// This should be adjusted to be an integer.
+constexpr float kStepSecs = 0.01161;
+// results in 43 Hz @ 44.1 kHz / 47 Hz @ 48 kHz / 47 Hz @ 96 kHz
+constexpr int kMaximumBinSizeHz = 50; // Hz
+// This is a quick hack to make a beatmap with only downbeats - will affect the bpm
+constexpr bool useDownbeatOnly = true;
+
+DFConfig makeDetectionFunctionConfig(int stepSize, int windowSize) {
+    // These are the defaults for the VAMP beat tracker plugin we used in Mixxx
+    // 2.0.
+    DFConfig config;
+    config.DFType = DF_COMPLEXSD;
+    config.stepSize = stepSize;
+    config.frameLength = windowSize;
+    config.dbRise = 3;
+    config.adaptiveWhitening = 0;
+    config.whiteningRelaxCoeff = -1;
+    config.whiteningFloor = -1;
+    return config;
+}
+} // namespace
+
+
+AnalyzerRhythm::AnalyzerRhythm(UserSettingsPointer pConfig)
+        : m_iSampleRate(0),
+          m_iTotalSamples(0),
+          m_iMaxSamplesToProcess(0),
+          m_iCurrentSample(0),
+          m_iMinBpm(0),
+          m_iMaxBpm(9999) {
+}
+
+bool AnalyzerRhythm::initialize(TrackPointer pTrack, int sampleRate, int totalSamples) {
+    if (totalSamples == 0) {
+        return false;
+    }
+
+    bool bpmLock = pTrack->isBpmLocked();
+    if (bpmLock) {
+        qDebug() << "Track is BpmLocked: Beat calculation will not start";
+        return false;
+    }
+
+    m_iSampleRate = sampleRate;
+    m_iTotalSamples = totalSamples;
+    m_iMaxSamplesToProcess = m_iTotalSamples;
+    m_iCurrentSample = 0;
+
+    m_stepSize = m_iSampleRate * kStepSecs;
```

Comment:
`m_stepSize` can be derived from `m_iSampleRate`.
Can we introduce a function `getStepSize() const` to calculate this on-demand rather than storing it in a member variable?

Reply:
Minimizing mutable state is a good idea.